tests/asg/asg_test.go

```diff
@@ -0,0 +1,156 @@
+
+	// https://github.com/kubernetes/client-go/issues/242
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const (
+	scaleTimeout = 10 * time.Minute
+)
+
+func TestASG(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	var specReporters []Reporter
+	junitReporter := reporters.NewJUnitReporter("/testresults/junit_basic.xml")
+	specReporters = append(specReporters, junitReporter)
+	RunSpecsWithDefaultAndCustomReporters(t, "Torpedo : ASG", specReporters)
+}
+
+var _ = BeforeSuite(func() {
+	InitInstance()
+})
+
+// This test performs basic test of scaling up and down the asg cluster
+var _ = Describe("{ClusterScaleUpDown}", func() {
+	It("has to validate that storage nodes are not lost during asg scaledown", func() {
+
+		autoNodeRecoveryTimeout := os.Getenv("ASG_AUTO_NODE_RECOVERY_TIMEOUT_MIN")
+		if autoNodeRecoveryTimeout == "" {
+			// Default storage nodes recovery timeout of 15 minutes
+			autoNodeRecoveryTimeout = "15"
+		}
+
+		autoNodeRecoveryTimeoutMins, err := strconv.Atoi(autoNodeRecoveryTimeout)
+		Expect(err).NotTo(HaveOccurred())
+
+		var contexts []*scheduler.Context
+		for i := 0; i < Inst().ScaleFactor; i++ {
+			contexts = append(contexts, ScheduleAndValidate(fmt.Sprintf("asgscaleupdown-%d", i))...)
+		}
+
+		intitialNodeCount, err := Inst().N.GetASGClusterSize()
+		Expect(err).NotTo(HaveOccurred())
+
+		scaleupCount := intitialNodeCount + intitialNodeCount/2
+		Step(fmt.Sprintf("scale up cluster from %d to %d nodes and validate",
+			intitialNodeCount, (scaleupCount/3)*3), func() {
+
+			Scale(scaleupCount)
+			Step(fmt.Sprintf("validate number of storage nodes after scale up"), func() {
+				Validate(scaleupCount)
+			})
+
+		})
+
+		Step(fmt.Sprintf("scale down cluster back to original size of %d nodes",
+			intitialNodeCount), func() {
+			Scale(intitialNodeCount)
+
+			Step(fmt.Sprintf("wait for %v minutes for auto recovery of storeage nodes",
+				autoNodeRecoveryTimeout), func() {
+				time.Sleep(time.Duration(autoNodeRecoveryTimeoutMins) * time.Minute)
+			})
+
+			// After scale down, get fresh list of nodes
+			// by re-initializing scheduler and volume driver
+			err = Inst().S.RefreshNodeRegistry()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = Inst().V.RefreshDriverEndpoints()
+			Expect(err).NotTo(HaveOccurred())
+
+			Step(fmt.Sprintf("validate number of storage nodes after scale down"), func() {
+				Validate(intitialNodeCount)
+			})
+		})
+
+		opts := make(map[string]bool)
+		opts[scheduler.OptionsWaitForResourceLeakCleanup] = true
+		ValidateAndDestroy(contexts, opts)
+
+	})
+})
+
+var _ = AfterSuite(func() {
+	PerformSystemCheck()
+	CollectSupport()
+	ValidateCleanup()
+})
+
+func init() {
+	ParseFlags()
+}
+
+func Scale(newCount int64) {
+	// In multi-zone ASG cluster, node count is per zone
+	perZoneNewCount := newCount / 3
+
+	err := Inst().N.SetASGClusterSize(perZoneNewCount, scaleTimeout)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func Validate(newCount int64) {
```

Comment:
name `newCount` to just `count` since for a `Validate` function, it's just the count that matters and not whether it's new.

rename `Validate` to `ValidateClusterSize`.